### PR TITLE
Add plugins feature

### LINKED ISSUE #323: More effective fuzzing
I got a question concerning this afl.rs implementation and reviewed the setup.

What I gathered is that basically afl-fuzz is compiled, but none of the llvm plugins, and instead the vanilla llvm features are used (specifically sancov).
This makes the integration easy, on the other hand cool features are then not available in the project, namely cmplog, auto dictionary/dict2file and faster instrumentation.

a mini cmplog could be done natively by adding `sanitizer-coverage-trace-compares`  (and perhaps trace-div, trace-gep, trace-loads, trace-stores, but unsure about the impact on effectiveness positively or negatively) here:
https://github.com/rust-fuzz/afl.rs/blob/ab62d59ba41c2f101c53e211746f7412ee652745/src/bin/cargo-afl.rs#L365
and then handing `afl-fuzz` the additional parameters with`-c 0` when starting as `cargo afl fuzz  ...`

A better but more complex approach would be to use the rust llvm framework to compile the afl++ llvm plugins and load these.
that would result in full cmplog, faster instrumentation, auto dictionary/dict2file, switch case splitting etc.

### PR DESCRIPTION
**Pull request by brunoproduit**
This PR activates full cmplog by compiling the afl++ llvm plugins and linking them into the target binaries.
This is activated via an added features "cmplog" as it needs nightly and llvm tools. It is not activated by default.

This fixes #323 .

### CODE REVIEW

**Comment by brunoproduit** [comment]
Pipelines seems to break at `afl system-config` and should probably be running `cargo install --path .` beforehand

**Comment by smoelius** [comment]
Thank you very much for this PR.

It's probably going to take a few rounds of review for me to understand all of what has changed, so please be patient with me.

I will start on this soon.

File: cargo-afl/src/bin/cargo-afl.rs [hunk 1]
```diff
@@ -1,3 +1,5 @@
+#![allow(unused_assignments)]
```
Comment (smoelius) [inline cargo-afl/src/bin/cargo-afl.rs:1]: Why is this needed?

File: cargo-afl/src/bin/cargo-afl.rs [hunk 2]
```diff
@@ -297,18 +299,49 @@ where
...
+             -C target-cpu=native "
+        );
+    } else {
+        rustflags = "-C llvm-args=-sanitizer-coverage-level=3 \
+           -C llvm-args=-sanitizer-coverage-trace-pc-guard \
+           -C llvm-args=-sanitizer-coverage-prune-blocks=0 \
+           -C llvm-args=-sanitizer-coverage-trace-compares
+           "
+        .to_string();
+    }
```
Comment (smoelius) [inline cargo-afl/src/bin/cargo-afl.rs:344]: I think the else branch is missing some flags.

I think it should be okay to reorder the flags. I would set more of them prior to entering this `if` condition, and then use the `if` condition to only set the ones that differ when `cmplog` is set/unset.

File: cargo-afl/src/bin/cargo-afl.rs [hunk 3]
```diff
@@ -343,6 +376,9 @@ where
         .env("RUSTDOCFLAGS", &rustdocflags)
         .env("ASAN_OPTIONS", asan_options)
         .env("TSAN_OPTIONS", tsan_options)
+        .env("AFL_LLVM_INSTRUMENT", "PCGUARD")
+        .env("AFL_LLVM_CMPLOG", "1")
+        .env("AFL_QUIET", "1")
```
Comment (smoelius) [inline cargo-afl/src/bin/cargo-afl.rs:381]: Shouldn't these be guarded by the `cmplog` feature?

If that's correct, then I would set these in the same `if` block above where the `-Z` flags were added.

**Review by smoelius — CHANGES_REQUESTED** [review]
Thanks a lot for doing this. I like the approach of using a Cargo feature.

Again, I may need to look at this a few times. So thanks in advance for your patience.

**Comment by brunoproduit** [comment]
Thanks a lot for the review! I'll take a look at it next week and make the requested changes

**Comment by vanhauser-thc** [comment]
Some context why this is something someone would want: it allows solving eg

For context for the cmplog, this is what AFL++ can solve:
 

                let slice16 = &data[20..36];
                let if_u128 = u128::from_le_bytes(slice16.try_into().unwrap());
                if secret_u128 != if_u128 { return; }

Plus the string matching you see added in the example plus floating point solving - all things that libfuzzer and honggfuzz can’t do.

File: cargo-afl/src/bin/cargo-afl.rs (hunk 1 above)
Comment (brunoproduit) [reply cargo-afl/src/bin/cargo-afl.rs]: Unnecessary and removed

File: cargo-afl/build.rs [hunk 4]
```diff
@@ -14,67 +13,90 @@ static AR_CMD: &str = "ar";
...
+        // skip the checks for the legacy x86 afl-gcc compiler
+        .env("AFL_NO_X86", "1")
+        // build just the runtime to avoid troubles with Xcode clang on macOS
+        //.env("NO_BUILD", "1")
+        .env("DESTDIR", common::afl_dir(base))
+        .env("PREFIX", "")
+        .env("LLVM_CONFIG", llvm_config)
+        .env_remove("DEBUG");
+    let status = command.status().expect("could not run 'make install'");
     assert!(status.success());
```
[the remark this answers is not in the document]
Comment (brunoproduit) [reply cargo-afl/build.rs:84]: I think I talked with @vanhauser-thc about this and those options were unnecessary and we removed them. The added the clean later on, which was also unnecessary. Actually if you `cargo install --git https://github.com/rust-fuzz/afl.rs` it will not work, as cargo doesn't handle redirection properly in the submodules of AFL++. I tried working around it with doing the submodule init in the buildscript.

File: cargo-afl/src/common.rs [hunk 5]
```diff
@@ -57,18 +57,30 @@ pub fn afl_dir(base: Option<&Path>) -> PathBuf {
...
+#[allow(dead_code)]
+#[must_use]
+pub fn get_llvm_config() -> String {
+    // Fetch the llvm version of the rust toolchain and set the LLVM_CONFIG environement variable to the same version
+    // This is needed to compile the llvm plugins (needed for cmplog) from afl with the right LLVM version
+    let version_meta = rustc_version::version_meta().unwrap();
+    let llvm_version = version_meta.llvm_version.unwrap().major.to_string();
+    let mut llvm_config = "llvm-config-".to_string();
+    llvm_config.push_str(&llvm_version);
+    llvm_config
```
[the remark this answers is not in the document]
Comment (brunoproduit) [reply cargo-afl/src/common.rs:85]: Changed in the last commit

File: cargo-afl/build.rs [hunk 6]
```diff
@@ -14,67 +13,90 @@ static AR_CMD: &str = "ar";
...
+        .status()
+        .expect("could not run 'git submodule init'");
+    assert!(status.success());
+
+    let mut command = Command::new("git");
+    command.args(["submodule", "update"]);
+
+    let status = command
+        .status()
+        .expect("could not run 'git submodule update'");
```
[the remark this answers is not in the document]
Comment (brunoproduit) [reply cargo-afl/build.rs]: See my other comment on line 99

File: afl/examples/cmplog.rs [hunk 7]
```diff
@@ -26,6 +26,13 @@ fn main() {
             return;
         };
 
+        let slice = &data[16..27];
+        let match_string = "HelloWorld";
```
[the remark this answers is not in the document]
Comment (brunoproduit) [reply afl/examples/cmplog.rs]: Changed in last commit

File: afl/examples/cmplog.rs [hunk 8]
```diff
@@ -26,6 +26,13 @@ fn main() {
             return;
         };
 
+        let slice = &data[16..27];
```
[the remark this answers is not in the document]
Comment (brunoproduit) [reply afl/examples/cmplog.rs]: Changed in last commit

File: cargo-afl/src/common.rs [hunk 9]
```diff
@@ -57,18 +57,30 @@ pub fn afl_dir(base: Option<&Path>) -> PathBuf {
 
 #[allow(dead_code)]
 #[must_use]
-pub fn afl_llvm_rt_dir(base: Option<&Path>) -> PathBuf {
-    data_dir(base, "afl-llvm-rt")
+pub fn afl_llvm_dir(base: Option<&Path>) -> PathBuf {
```
[the remark this answers is not in the document]
Comment (brunoproduit) [reply cargo-afl/src/common.rs:60]: Exactly, in the build script all shared libraries are copied in the same folder

File: cargo-afl/build.rs [hunk 10]
```diff
@@ -14,67 +13,90 @@ static AR_CMD: &str = "ar";
...
+
+    if cfg!(feature = "cmplog") {
+        llvm_config = common::get_llvm_config();
+    }
+
+    build_afl(&work_dir, base, llvm_config);
+
+    if cfg!(feature = "cmplog") {
+        build_afl_llvm_runtime(&work_dir, base);
+    }
```
[the remark this answers is not in the document]
Comment (brunoproduit) [reply cargo-afl/build.rs:34]: Correct, I wonder what I was trying to do here. Let me change this by creating a new function

**Comment by brunoproduit** [comment]
Did a first round of fixes for most of the conversations, let me know if I missed something.

File: cargo-afl/build.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (smoelius) [reply cargo-afl/build.rs:85]: > `cargo install --git https://github.com/rust-fuzz/afl.rs`

Is that why the PR involves submodules? To enable above command?

If so, is enabling the above command needed to enable the `cmplog` feature? (It feels to me like it's orthogonal.)

[the remark this answers is not in the document]
Comment (brunoproduit) [reply cargo-afl/build.rs:85]: Not exactly, I tried using `cargo install --git https://github.com/rust-fuzz/afl.rs` which didn't work as the submodule didn't redirect properly. Tried hotpatching this in the buildscript, which was a bad idea and forgot to remove it. I removed it in the last commit

File: cargo-afl/src/bin/cargo-afl.rs [hunk 11]
```diff
@@ -297,18 +298,59 @@ where
...
+            .unwrap_or_else(|_| panic!("could not run {llvm_config} --version"));
+        assert!(status.success());
+
+        rustflags.push_str(&format!(
+            "-Z llvm-plugins={p}/cmplog-instructions-pass.so  \
+            -Z llvm-plugins={p}/cmplog-routines-pass.so \
+            -Z llvm-plugins={p}/cmplog-switches-pass.so \
+            -Z llvm-plugins={p}/SanitizerCoveragePCGUARD.so
+             "
+        ));
```
Comment (smoelius) [inline cargo-afl/src/bin/cargo-afl.rs:340]: For my understanding, why is this list of `so`s less than those listed in `build_afl_llvm_plugins`?

File: cargo-afl/build.rs [hunk 12]
```diff
@@ -94,6 +101,33 @@ fn build_afl_llvm_runtime(work_dir: &Path, base: Option<&Path>) {
     assert!(status.success());
 }
 
+fn build_afl_llvm_plugins(work_dir: &Path, base: Option<&Path>) {
```
Comment (smoelius) [inline cargo-afl/build.rs:104]:
```suggestion
fn copy_afl_llvm_plugins(work_dir: &Path, base: Option<&Path>) {
```
Nit, since the pluigns were actually built in `build_afl`.

File: cargo-afl/src/bin/cargo-afl.rs [hunk 13]
```diff
@@ -297,18 +298,59 @@ where
...
+
+        environment_variables.insert("AFL_LLVM_INSTRUMENT".to_string(), "PCGUARD".to_string());
+        environment_variables.insert("AFL_LLVM_CMPLOG".to_string(), "1".to_string());
+        environment_variables.insert("AFL_QUIET".to_string(), "1".to_string());
```
Comment (smoelius) [inline cargo-afl/src/bin/cargo-afl.rs:344]: Why do you set `AFL_QUIET`?

File: cargo-afl/src/bin/cargo-afl.rs [hunk 14]
```diff
@@ -297,18 +298,59 @@ where
...
+            .status()
+            .unwrap_or_else(|_| panic!("could not run {llvm_config} --version"));
+        assert!(status.success());
+
+        rustflags.push_str(&format!(
+            "-Z llvm-plugins={p}/cmplog-instructions-pass.so  \
+            -Z llvm-plugins={p}/cmplog-routines-pass.so \
+            -Z llvm-plugins={p}/cmplog-switches-pass.so \
+            -Z llvm-plugins={p}/SanitizerCoveragePCGUARD.so
+             "
```
Comment (smoelius) [inline cargo-afl/src/bin/cargo-afl.rs:339]:
```suggestion
            "
```
Nit.

**Review by smoelius — CHANGES_REQUESTED** [review]

**Comment by smoelius** [comment]
Thanks for your continued patience, @brunoproduit.

File: cargo-afl/src/bin/cargo-afl.rs (hunk 11 above)
Comment (brunoproduit) [reply cargo-afl/src/bin/cargo-afl.rs:340]: We were thinking to add some functionality in the future that depends on these other plugins, thus included all of them in the copied directory

File: cargo-afl/build.rs [hunk 15]
```diff
@@ -14,67 +13,75 @@ static AR_CMD: &str = "ar";
...
     #[cfg(unix)]
     let _file = sys::lock_path(&work_dir).unwrap();
 
-    build_afl(&work_dir, base.as_deref());
-    build_afl_llvm_runtime(&work_dir, base.as_deref());
+    let mut llvm_config = String::new();
+
+    if cfg!(feature = "cmplog") {
+        llvm_config = common::get_llvm_config();
+    }
```
[the remark this answers is not in the document]
Comment (brunoproduit) [reply cargo-afl/build.rs]: Done in the latest commit

File: cargo-afl/build.rs [hunk 16]
```diff
@@ -14,67 +13,75 @@ static AR_CMD: &str = "ar";
 mod common;
 
 fn main() {
```
[the remark this answers is not in the document]
Comment (brunoproduit) [reply cargo-afl/build.rs]: This was initially because of the compilation errors we had and discussed before, all these have been reverted to the master branch's way. Done in the latest commit.

File: cargo-afl/build.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (brunoproduit) [reply cargo-afl/build.rs:105]: Recombined them, but removed the unnecessary "all"

File: cargo-afl/build.rs [hunk 17]
```diff
@@ -94,6 +101,33 @@ fn build_afl_llvm_runtime(work_dir: &Path, base: Option<&Path>) {
...
+            .unwrap_or_else(|_| panic!("Couldn't copy shared object file {sl}"));
+    }
+
+    let status = Command::new(AR_CMD)
+        .arg("r")
+        .arg(common::archive_file_path(base))
+        .arg(common::object_file_path(base))
+        .status()
+        .expect("could not run 'ar'");
+    assert!(status.success());
```
[the remark this answers is not in the document]
Comment (brunoproduit) [reply cargo-afl/build.rs]: Indeed, good catch, removed

File: cargo-afl/build.rs (hunk 12 above)
Comment (brunoproduit) [reply cargo-afl/build.rs]: Done in the latest commit

File: cargo-afl/src/bin/cargo-afl.rs (hunk 14 above)
Comment (brunoproduit) [reply cargo-afl/src/bin/cargo-afl.rs]: Done in the latest commit

File: cargo-afl/src/bin/cargo-afl.rs [hunk 18]
```diff
@@ -297,18 +298,59 @@ where
...
+
+        let llvm_config = common::get_llvm_config();
+
+        // check if llvm tools are installed and with the good version for the plugin compilation
+        let mut command = Command::new(llvm_config.clone());
+        command.args(["--version"]);
+        let status = command
+            .status()
+            .unwrap_or_else(|_| panic!("could not run {llvm_config} --version"));
+        assert!(status.success());
```
[the remark this answers is not in the document]
Comment (brunoproduit) [reply cargo-afl/src/bin/cargo-afl.rs:332]: As you prefer, we wanted to make sure that the environment is the same when using it, let me remove that

**Comment by brunoproduit** [comment]
No problem! @smoelius If you could take yet another look, I should have resolved all the new conversations

**Comment by smoelius** [comment]
I'm likely going to request more changes, but I want to be sure this is going to work in CI.

So, I'm going to push to your branch. But I won't touch any files that you've modified. Ok?

**Comment by smoelius** [comment]
I need to find the right syntax for the GitHub matrix and macOS's sed. I'll get back to this soon.

**Comment by smoelius** [comment]
When I `brew install llvm@17` on a Mac, I get an `llvm-config` but not an `llvm-config-17`.

Mac is a platform that we support, and I expect many Mac users will install llvm with Homebrew.

Could you please adjust the build script so that `llvm-config-17` is not required (at least not on macOS)?

**Comment by brunoproduit** [comment]
We do need to check that the major version matches the rustc llvm version, else it won't work. I'll set an if that sets `LLVM_CONFIG` to "llvm-config" on Mac and check that the major version corresponds (by catching the output of the --version). later this week if that works for you

**Comment by smoelius** [comment]
> I'll set an if that sets `LLVM_CONFIG` to "llvm-config" on Mac and check that the major version corresponds (by catching the output of the --version). later this week if that works for you

That sounds perfect. Thank you!

**Comment by smoelius** [comment]
I'm going to push more changes to rust.yml, but I will stay out of your code.

**Comment by vanhauser-thc** [comment]
btw I added support for two advanced features:
  - dict2file: generates a valid dictionary on the fly during compilation (needs AFL_LLVM_DICT2FILE=/file/to/create.txt set), this has shown on fuzzbench to be a major improvement to unlock coverage
  -  selective coverage: only instrument for fuzzing specific functions or files for fuzzing - or specifically not instrument for fuzzing.
 
also in tests I see that using the AFL++ instrumentation module vs. the llvm sancov creates a x2 speed increase (without any drawbacks and an even better coverage gathering).

**Comment by smoelius** [comment]
@brunoproduit I'm pushing to your branch to get the fix for https://github.com/AFLplusplus/AFLplusplus/issues/1898.

**Comment by brunoproduit** [comment]
Should we change the feature name to something else now that this activates more than cmplog?
What about something like "llvm-afl-plugins"?

**Comment by brunoproduit** [comment]
Features are 
- Cmplog
- dict2file
- partial instrumentation
- Faster fuzzing due to the AFL++ instrumentation module instead of the llvm sancov

IMO, we shouldn't separate as they all have the same requirements. Either you want to add addons from AFL, or not.
The only thing is you need nightly and LLVM, which is why a feature is nice.

Haven't had time to take a look at the macos issue yet, sorry

**Comment by smoelius** [comment]
> Features are
> 
> * Cmplog
> * dict2file
> * partial instrumentation
> * Faster fuzzing due to the AFL++ instrumentation module instead of the llvm sancov

Thanks.

> IMO, we shouldn't separate as they all have the same requirements. Either you want to add addons from AFL, or not. The only thing is you need nightly and LLVM, which is why a feature is nice.

I'm not sure I agree, but we don't have to decide this now. I agree that "llvm-afl-plugins" would be a better feature name.

**Comment by vanhauser-thc** [comment]
Can we please shorten the flag to “plugins”? Imho it is too long otherwise

**Comment by smoelius** [comment]
> Can we please shorten the flag to “plugins”? Imho it is too long otherwise

That works for me.

**Comment by smoelius** [comment]
> Separately, I am trying to diagnose why the shared libraries cannot be loaded on macOS ([see here](https://github.com/rust-fuzz/afl.rs/actions/runs/6651499455/job/18073649214?pr=392#step:7:51)). From what I can tell, it's not a path problem, i.e., I expect that `/Users/runner/.../cmplog-instructions-pass.so` exists, as I observe the same failure locally.

This is the error I get locally:
> Could not load library '/Users/[snip]/cmplog-instructions-pass.so': dlopen(/Users/[snip]/cmplog-instructions-pass.so, 0x0009): symbol not found in flat namespace '__ZN4llvm17PreservedAnalyses14AllAnalysesKeyE'

**Comment by smoelius** [comment]
@vanhauser-thc Can I ask why AFLplusplus's macos-latest tests [are disabled](https://github.com/AFLplusplus/AFLplusplus/blob/e241280fcc0d26066d976495251ca79a3f37509f/.github/workflows/ci.yml#L39-L58)?

**Comment by vanhauser-thc** [comment]
> @vanhauser-thc Can I ask why AFLplusplus's macos-latest tests [are disabled](https://github.com/AFLplusplus/AFLplusplus/blob/e241280fcc0d26066d976495251ca79a3f37509f/.github/workflows/ci.yml#L39-L58)?

Because there are often unpredictable changes. Eg Apple change how to preload libraries and supporting this would be quite an effort. Or the he crippled Xcode clang. And brew llvm also changes from time to time how it behaves. Finally GitHub’s mac container also changes sometimes where otherwise it works perfectly on my Mac mini.
So making the CI work is quite an effort.
I am not a Mac user and few people fuzz on Mac (bad performance of clone for example) hence I rather fix stuff it I get a report something is not working.

**Comment by smoelius** [comment]
In `GNUmakefile.llvm`, if I change [this line](https://github.com/AFLplusplus/AFLplusplus/blob/85c5b5218c6a7b2289f309fbd1625a5d0a602a00/GNUmakefile.llvm#L314):
```sh
  CLANG_LFL += -Wl,-flat_namespace -Wl,-undefined,suppress
```
to this one:
```sh
  CLANG_LFL += -Wl,-flat_namespace -Wl,-undefined,suppress `$(LLVM_CONFIG) --libs`
```
I get a different error:
> Assertion failed: (AnalysisPasses.count(PassT::ID()) && "This analysis pass was not registered prior to being queried"), function getResult, file PassManager.h, line 776.

Does this mean anything to you? @brunoproduit @vanhauser-thc

**Comment by vanhauser-thc** [comment]
MacOS is difficult. brew needs to install the right llvm version, there all the time changes in brew and Mac - for me as a non-user of MacOS it is very difficult to be on top of this.
Hence there is no full MacOS support for AFL++ and only act on things if people report issues.

I would simply propose to not support the plugin feature for MacOS, someone can do it by hand if they want to,.
Handling all the CI failures that will creep up again and again over the time is not worth it IMHO

**Comment by vanhauser-thc** [comment]
@smoelius please note that I changed the output of the --version command. IHMO it is important to see if afl.rs was compiled normal or with the plugins feature. if you want it to look different, just edit away or tell me how else you would like it :)

**Review by smoelius — COMMENTED** [review]
I think once these changes are made, this should be mergeable.

File: cargo-afl/src/bin/cargo-afl.rs [hunk 19]
```diff
@@ -89,20 +88,32 @@ fn main() {
...
+    ($s1:expr, $s2:expr) => {
+        concat!($s1, $s2)
+    };
+}
+
 #[allow(clippy::too_many_lines)]
 fn clap_app() -> clap::Command {
     use clap::{value_parser, Arg, Command};
 
     let help = "In addition to the subcommands above, Cargo subcommands are also \
```
[the remark this answers is not in the document]
Comment (brunoproduit) [reply cargo-afl/src/bin/cargo-afl.rs]: Handled in https://github.com/brunoproduit/afl.rs/commit/a6716abf6d03e946b3b95b818ead35e31bcb2dc6

File: cargo-afl/src/bin/cargo-afl.rs [hunk 20]
```diff
@@ -89,20 +88,32 @@ fn main() {
...
                       supported (see `cargo help` for a list of all Cargo subcommands).";
 
+    const VERSION: &'static str = if cfg!(feature = "plugins") {
```
[the remark this answers is not in the document]
Comment (brunoproduit) [reply cargo-afl/src/bin/cargo-afl.rs]: Handled in https://github.com/brunoproduit/afl.rs/commit/a6716abf6d03e946b3b95b818ead35e31bcb2dc6

File: cargo-afl/build.rs [hunk 21]
```diff
@@ -60,21 +61,41 @@ fn main() {
...
+    }
 }
 
 fn build_afl(work_dir: &Path, base: Option<&Path>) {
+    let mut environment_variables = HashMap::<String, String>::new();
+
+    if cfg!(feature = "plugins") {
+        let llvm_config = check_llvm_and_get_config();
+        environment_variables.insert("LLVM_CONFIG".to_string(), llvm_config);
+    }
```
[the remark this answers is not in the document]
Comment (brunoproduit) [reply cargo-afl/build.rs]: Handled in https://github.com/brunoproduit/afl.rs/commit/a6716abf6d03e946b3b95b818ead35e31bcb2dc6

File: cargo-afl/build.rs [hunk 22]
```diff
@@ -60,21 +61,41 @@ fn main() {
...
-        .env("DESTDIR", common::afl_dir(base))
-        .env("PREFIX", "")
+        .args(["clean", "install"])
+        .envs(&environment_variables)
         .env_remove("DEBUG");
-    let status = command.status().expect("could not run 'make'");
+
+    let status = command
+        .status()
+        .expect("could not run 'make clean, make install'");
```
[the remark this answers is not in the document]
Comment (brunoproduit) [reply cargo-afl/build.rs]: Handled in https://github.com/brunoproduit/afl.rs/commit/a6716abf6d03e946b3b95b818ead35e31bcb2dc6

File: cargo-afl/build.rs [hunk 23]
```diff
@@ -94,6 +115,60 @@ fn build_afl_llvm_runtime(work_dir: &Path, base: Option<&Path>) {
...
+        "afl-llvm-dict2file.so",
+        "afl-llvm-pass.so",
+        "cmplog-instructions-pass.so",
+        "cmplog-routines-pass.so",
+        "cmplog-switches-pass.so",
+        "compare-transform-pass.so",
+        "split-compares-pass.so",
+        "split-switches-pass.so",
+        "SanitizerCoveragePCGUARD.so",
+    ];
```
[the remark this answers is not in the document]
Comment (brunoproduit) [reply cargo-afl/build.rs]: The idea was to fix those for readability and future-proofing, like if we want to exclude certain so's in the future. I changed this in the last commit

File: cargo-afl/src/bin/cargo-afl.rs [hunk 24]
```diff
@@ -89,20 +88,32 @@ fn main() {
...
+        concat_str!(env!("CARGO_PKG_VERSION"), " [feature=plugins]")
```
[the remark this answers is not in the document]
Comment (brunoproduit) [reply cargo-afl/src/bin/cargo-afl.rs]: fixed in the last commit

File: cargo-afl/build.rs [hunk 25]
```diff
@@ -60,21 +61,41 @@ fn main() {
...
     build_afl(&work_dir, base.as_deref());
     build_afl_llvm_runtime(&work_dir, base.as_deref());
+
+    if cfg!(feature = "plugins") {
+        copy_afl_llvm_plugins(&work_dir, base.as_deref());
+    }
 }
 
 fn build_afl(work_dir: &Path, base: Option<&Path>) {
+    let mut environment_variables = HashMap::<String, String>::new();
```
[the remark this answers is not in the document]
Comment (brunoproduit) [reply cargo-afl/build.rs]: fixed in the last commit

File: cargo-afl/src/bin/cargo-afl.rs [hunk 26]
```diff
@@ -297,18 +308,48 @@ where
...
-         -C llvm-args=-sanitizer-coverage-trace-compares \
-         -C opt-level=3 \
-         -C target-cpu=native "
+             -C overflow_checks \
+             -C passes={passes} \
+             -C codegen-units=1 \
+             -C opt-level=3 \
+             -C target-cpu=native "
     );
+    let mut environment_variables = HashMap::<String, String>::new();
```
[the remark this answers is not in the document]
Comment (brunoproduit) [reply cargo-afl/src/bin/cargo-afl.rs]: fixed in the last commit

**Comment by brunoproduit** [comment]
Applied the requested changes in the last commit, could take a look @smoelius ?

**Comment by smoelius** [comment]
@brunoproduit Could you please resolve the conflicts?

If it's easiest, you can squash all of the commits down to one. (I would like to squash before rebasing anyway.)

**Comment by brunoproduit** [comment]
@smoelius done, you can go ahead and merge/squash it

**Review by smoelius — APPROVED** [review]

**Comment by smoelius** [comment]
Thank you, @brunoproduit.

**Comment by smoelius** [comment]
@vanhauser-thc @brunoproduit https://github.com/rust-fuzz/afl.rs/pull/421 impacts the feature introduced by this PR.

Could I impose on one or both of you try #421, and let me know if there are any suprises?

**Comment by brunoproduit** [comment]
I'll try it out this week, I like the idea

**Comment by jberryman** [comment]
> dict2file: generates a valid dictionary on the fly during compilation (needs AFL_LLVM_DICT2FILE=/file/to/create.txt set), this has shown on fuzzbench to be a major improvement to unlock coverage
selective coverage: only instrument for fuzzing specific functions or files for fuzzing - or specifically not instrument for fuzzing.

it would be super helpful if this stuff was in the README or somewhere prominent.

**Comment by kevin-valerio** [comment]
> > Separately, I am trying to diagnose why the shared libraries cannot be loaded on macOS ([see here](https://github.com/rust-fuzz/afl.rs/actions/runs/6651499455/job/18073649214?pr=392#step:7:51)). From what I can tell, it's not a path problem, i.e., I expect that `/Users/runner/.../cmplog-instructions-pass.so` exists, as I observe the same failure locally.
> 
> This is the error I get locally:
> 
> > Could not load library '/Users/[snip]/cmplog-instructions-pass.so': dlopen(/Users/[snip]/cmplog-instructions-pass.so, 0x0009): symbol not found in flat namespace '__ZN4llvm17PreservedAnalyses14AllAnalysesKeyE'

Hey. I'm having the same error here. How did you fix it, do you remember ?

**Comment by smoelius** [comment]
@kevin-valerio I never was able to resolve this issue. (See https://github.com/rust-fuzz/afl.rs/pull/392#issuecomment-1784033265.) That is why plugins are not tested on macOS in CI: https://github.com/rust-fuzz/afl.rs/blob/425b419dbee7b0c9bcfa2dbb6868bb4704972b9b/.github/workflows/rust.yml#L38-L39
I just checked and I still experience the issue. FWIW, I would love to have help with this.

**Comment by vanhauser-thc** [comment]
Are you sure the plugins are compile with the correct clang from brew? Xcode clang is crippled and the brew llvm version must match the version rust is based on (currently 18)

**Comment by kevin-valerio** [comment]
> Are you sure the plugins are compiled with the correct clang from brew? Xcode clang is crippled and the brew llvm version must match the version rust is based on (currently 18)
```
❯ rustc --version --verbose
rustc 1.81.0-nightly (6b0f4b5ec 2024-06-24)
LLVM version: 18.1.7

❯ brew info llvm
==> llvm: stable 18.1.7 (bottled), HEAD [keg-only]

❯ which clang
/opt/homebrew/opt/llvm/bin/clang

❯ which llvm-ar
/opt/homebrew/opt/llvm/bin/llvm-ar
```

Everything seem to match for me unfortunately

**Comment by vanhauser-thc** [comment]
Then likely brew updated llvm to a slightly newer minor revision at a later point so you need to recompile the plugins …

cargo afl config —plugins —force
